### ISSUE TITLE
Dependency on header files

### ISSUE THREAD

**Issue by carlobaldassi** [JuliaLang/BinDeps.jl#119]
On Linux/BSD, [GLPK.jl](https://github.com/JuliaOpt/GLPK.jl) builds libglpk from source, using an Autotool provider. The build requires the gmp.h headers. When Julia is installed from source, these are found in the Julia tree; however, when it is installed from @staticfloat's PPA under Ubuntu they are typically not available, since only the binary is installed, and the installation fails.

The problem is: I don't know how to detect whether the header is there, as it seems that `library_dependency` only cares about the binary. I tried some simple `validate` functions without success. Basically, I'd need to do the same thing which autotools' `configure` does when it looks for a header file.

When testing on Travis, we need to add an `apt-get install libgmp-dev` manually.

Did I just miss something obvious? Or is this the same problem as #103?

**Comment by carlobaldassi** [comment]
Yes, ideally, this would be a dependency of the libglpk `Source` or `Autotools` provider. In practical terms, it could be a generic dependency, since GLPK.jl always builds libglpk from source on Linux/BSD, and downloads a binary on the other platforms.
The real problem is detecting gmp.h, and disentangling it from libgmp.so. The current dependencies are - unless I'm missing something - defined in terms of the library binary. So when I tried to create an `AptGet` provider for `libgmp-dev` I ran into one of these two situations:

1. The package is installed but the dependency fails because it expects some `gmp-dev.so` file as a result (I think)
2. The package only checks for libgmp.so and therefore does not install libgmp-dev.

As I wrote above, I tried to use `validate` to fix the second case, but failed.
On top of this, an `AptGet` provider only works on Debian and derivatives, which is better then nothing but still not a complete solution, e.g. RedHat users would be left out.

A complete solution to the problem would probably require two additional mechanisms: one for specifying sub-dependencies, and one for checking header files.

**Comment by staticfloat** [comment]
@Keno how hard would it be to create a `BinDeps` target that is not a dynamic library?  As far as I can tell, all we really need is something akin to `library_dependency()` that is instead `file_dependency()` or somesuch.  You would have a better idea than I as to how much work it would be to sub in something that's not actually `dlopen()`-able.

**Comment by rdeits** [comment]
I have come across this issue as well when trying to build dependencies that rely on a header (like `Python.h`) or an executable (like `pkg-config` or `cmake`). I would love to be able to handle those as `BinDeps` targets.

**Comment by SylvainCorlay** [comment]
I am working around this limitation by producing a dummy binary along the header-only library.
